### ISSUE TITLE
Surrogate keys API seems broken with multiple sites

### ISSUE THREAD

**Issue by p0358** [darkweak/souin#446]
If you have multiple sites, it seems that `/souin-api/souin/surrogate_keys` will only list keys on the API site *opposite* to the first site a cachable request was made after server startup, as weird as that sounds. Then the site that served the first request would have `/souin-api/souin/surrogate_keys` always return `{}`, and PURGE wouldn't work on that either.

With that said, caching itself seems to be working properly, and puring keys on the other site clears them on the first one.

Expected behavior: `/souin-api/souin/surrogate_keys` would return the same output for all sites, since the cache storage is also shared (differentiated by hostname in main cache keys)

**Comment by p0358** [comment]
I've found the probable cause.

Firstly, new surrogateStorage is initialized here in NewHTTPCacheHandler func (and then assigned to SurrogateKeyStorer and also passed into GenerateHandlerMap, where the former is used later on for cache handling, and latter for API handling):
https://github.com/darkweak/souin/blob/5a88d84c26128d43e37c614c83c3d3e53da2afd1/pkg/middleware/middleware.go#L68-L97

Then, the Provision function creates NewHTTPCacheHandler, but at its end ensures that only single surrogate storage is used in the app:
https://github.com/darkweak/souin/blob/5a88d84c26128d43e37c614c83c3d3e53da2afd1/plugins/caddy/httpcache.go#L238-L290

Problem:
`SurrogateKeyStorer` gets changed, but what was passed to `InternalEndpointHandlers` does not!!

**Comment by p0358** [comment]
I confirm that doing this half-assed patch fixes the issue:
```go
if len(app.Storers) == 0 {
	app.Storers = s.SouinBaseHandler.Storers
} else {
	s.SouinBaseHandler.Storers = app.Storers
}

if app.SurrogateStorage == (surrogates_providers.SurrogateInterface)(nil) {
	app.SurrogateStorage = s.SouinBaseHandler.SurrogateKeyStorer
} else {
	s.SouinBaseHandler.SurrogateKeyStorer = app.SurrogateStorage
	s.SouinBaseHandler.InternalEndpointHandlers = api.GenerateHandlerMap(s.SouinBaseHandler.Configuration, s.SouinBaseHandler.Storers, app.SurrogateStorage)
}
```
I suppose it isn't really ideal to PR as-is, but it shows that the issue lied here at least

**Comment by darkweak** [comment]
Hello @p0358 a PR is available to use the storers (distributed or not) as storage for the surrogate-keys https://github.com/darkweak/souin/pull/448 because actually it was handled as in-memory map.

**Comment by p0358** [comment]
Hmm, but does it mean I could still have shared surrogate keys between two sites? Like if I purge by keys on one site, I want the relevant keys purged on all sites.

My use-case is having an external site with HTTPS and a bunch of rewrites, plus internal one just for ESI and cache API (but main site and ESI requests do share surrogate keys, and clearing them out does work after that janky proof-of-concept patch)

**Comment by darkweak** [comment]
If the responses are shared, when you purge, it will invalidate for both sites.

**Comment by Shotster** [comment]
I'm experiencing something similar but not sure if it's related. I _always_ get `{}` at the `surrogate_keys` endpoint even though I've defined a couple "catch-all" keys...

### Dynamic config snippet
```
http:
  services:
    my_svc:
      loadBalancer:
        servers:
          - url: https://my-domain.com
  routers:
    my_rtr:
      entrypoints: websecure
      rule: Host(`other-domain.com`) && PathPrefix(`/blog`)
      middlewares:
        #
        # middleware (compression, header/body rewrites)
        #
        - http-cache
      service: my_svc
      tls: {}
  middlewares:
    http-cache:
      plugin:
        souin:
          api:
            basepath: /souin-api
            souin:
              basepath: /souin
          default_cache:
            cache_name: mycache
            ttl: 3600s
          log_level: debug
          surrogate_keys:
            my_key_1:
              url: '.+'
            my_key_2:
              url: '.+'
```

Even more perplexing is that keys for the main page URLs never show up at `/souin-api/souin`. All the other keys - i.e. CSS, JS assets and XHR requests - do show up in the cache listing.

I know the cache is active based on performace as well as Chrome dev tools. For example, I can't see or purge the key shown below because it never appears in the cache listing...

```
Age:           327
Cache-Control: max-age=3600
Cache-Status:  mycache; hit; ttl=3273; key=GET-https-my-domain.com-/test/contact-us
```

Might this be fixed with [#448](https://github.com/darkweak/souin/pull/448), or should I submit a separate issue?

**Comment by Shotster** [comment]
Just following up to say that I'm now running the latest souin master branch (which appears to have incorporated the PR mentioned in this issue) in local mode on my dev machine, and I'm still experiencing the same behavior with one exception. Instead of the `{}` reponse on the `surrogate_keys` endpoint, I now get `null`. Aside from that, I'm still unable to see or purge top-level pages of my site despite them clearly being in the cache somewhere. 🤔
